### ISSUE TITLE
H5Location errors during 'make build-submodule'

### ISSUE THREAD

**Issue by mebbert** [PacificBiosciences/blasr#161]
Hi,

I'm trying to compile blasr, but I keep getting these errors:

    /apps/gcc/4.9.2/bin/g++ -std=c++11 -pedantic  -I../pbdata -I../hdf5-1.8.13-linux-x86_64-shared/include/     -c BufferedHDF2DArray.cpp -o BufferedHDF2DArray.o
    In file included from HDFGroup.hpp:9:0,
                 from HDFData.hpp:7,
                 from BufferedHDF2DArray.hpp:12,
                 from BufferedHDF2DArray.cpp:1:
    HDFAttributable.hpp:8:33: error: variable or field ‘CallStoreAttributeName’ declared void
    void CallStoreAttributeName(H5::H5Location &obj, std::string attrName, void *attrListPtr);
                                 ^
    HDFAttributable.hpp:8:29: error: ‘H5Location’ is not a member of ‘H5’
    void CallStoreAttributeName(H5::H5Location &obj, std::string attrName, void *attrListPtr);
                             ^
    HDFAttributable.hpp:8:45: error: ‘obj’ was not declared in this scope
    void CallStoreAttributeName(H5::H5Location &obj, std::string attrName, void *attrListPtr);
                                             ^
    HDFAttributable.hpp:8:62: error: expected primary-expression before ‘attrName’
    void CallStoreAttributeName(H5::H5Location &obj, std::string attrName, void *attrListPtr);
                                                              ^
    HDFAttributable.hpp:8:72: error: expected primary-expression before ‘void’
    void CallStoreAttributeName(H5::H5Location &obj, std::string attrName, void *attrListPtr);
                                                                        ^
    HDFAttributable.hpp:14:34: error: ‘H5::H5Location’ has not been declared
     void StoreAttributeNames(H5::H5Location &thisobject,
                                  ^
    HDFAttributable.hpp:17:17: error: ‘H5Location’ in namespace ‘H5’ does not name a type
     virtual H5::H5Location* GetObject();
                 ^
    In file included from HDFData.hpp:7:0,
                 from BufferedHDF2DArray.hpp:12,
                 from BufferedHDF2DArray.cpp:1:
    HDFGroup.hpp:24:9: error: ‘H5Location’ in namespace ‘H5’ does not name a type
     H5::H5Location* GetObject();
         ^
    In file included from BufferedHDF2DArray.hpp:12:0,
                 from BufferedHDF2DArray.cpp:1:
    HDFData.hpp:21:9: error: ‘H5Location’ in namespace ‘H5’ does not name a type
     H5::H5Location* GetObject();

Here is my configure line:

    ./configure.py --shared --sub --no-pbbam HDF5_INCLUDE=../hdf5-1.8.13-linux-x86_64-shared/include/ HDF5_LIB=../hdf5-1.8.13-linux-x86_64-shared/lib/

I then ran:
   
    make init-submodule

I tried using both pre-compiled and compiled-from-scratch versions of hdf5 1.8.13 (also tried 1.8.16). All versions were compiled with c++. I also tried both shared and static versions of the pre-compiled versions. Here are the files in ../hdf5-1.8.13-linux-x86_64-shared/lib/:

    libhdf5.a         libhdf5_cpp.so.8.0.2  libhdf5_fortran.so.8.0.2  libhdf5_hl_cpp.so.8      libhdf5hl_fortran.so.8      libhdf5_hl.so.8.0.2  libhdf5.so.8.0.2  libz.so.1
    libhdf5_cpp.a     libhdf5_fortran.a     libhdf5_hl.a              libhdf5_hl_cpp.so.8.0.2  libhdf5hl_fortran.so.8.0.2  libhdf5.la           libsz.so          libz.so.1.2.5
    libhdf5_cpp.la    libhdf5_fortran.la    libhdf5_hl_cpp.a          libhdf5hl_fortran.a      libhdf5_hl.la               libhdf5.settings     libsz.so.2
    libhdf5_cpp.so    libhdf5_fortran.so    libhdf5_hl_cpp.la         libhdf5hl_fortran.la     libhdf5_hl.so               libhdf5.so           libsz.so.2.0.0
    libhdf5_cpp.so.8  libhdf5_fortran.so.8  libhdf5_hl_cpp.so         libhdf5hl_fortran.so     libhdf5_hl.so.8             libhdf5.so.8         libz.so

I don't know what I'm missing. I've also tried nuking the blasr folder and starting over. I'm on Centos 6.6. Any ideas?

Really appreciate it.

**Comment by vrainish-pacbio** [comment]
Hi Mark !

Below is a step-by-step example of building Blasr from scratch (using prebuilt HDF5 libraries).
I’ve verified it for CentOS 7 and it works without the hitch.

Please, let me know what’s the your results ( I understand you are using CentOS 6)

Vladimir

P.S. I am about to put this guide to Blasr wiki today

------------------------------------------------------------------

Example of building blasr from scratch in /home/pacbio-user

Please, denote that blasr requires HDF5 library installed.
The directory for HDF5 in the example below is /home/pacbio-user/hdf5.
The directory for blasr itself is /home/pacbio-user/blasr

-> pwd
/home/pacbio-user

Load and untar prebuilt HDF5 libraries

-> mkdir hdf5
-> cd hdf5
-> wget http://www.hdfgroup.org/ftp/HDF5/current/bin/linux-centos6-x86_64-gcc447/hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared.tar.gz
-> tar -zxvf hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared.tar.gz

Clone blasr from github

-> cd ..
-> git clone https://github.com/PacificBiosciences/blasr.git --recursive
-> cd blasr

Configure and buid blasr

-> ./configure.py --shared --sub --no-pbbam  HDF5_INC=/home/pacbio-user/hdf5/hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared/include/ HDF5_LIB=/home/pacbio-user/hdf5/hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared/lib/
-> make configure-submodule
-> make build-submodule
-> make blasr
…
LD_LIBRARY_PATH=/home/pacbio-user/blasr/libcpp/alignment:/home/pacbio-user/blasr/libcpp/hdf:/home/pacbio-user/blasr/libcpp/pbdata:/home/pacbio-user/hdf5/hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared/lib/::::
…

-> export LD_LIBRARY_PATH=/home/pacbio-user/blasr/libcpp/alignment:/home/pacbio-user/blasr/libcpp/hdf:/home/pacbio-user/blasr/libcpp/pbdata:/home/pacbio-user/hdf5/hdf5-1.8.16-linux-centos6-x86_64-gcc447-shared/lib/::::

blasr executable is /home/pacbio-user/blasr/blasr

**Comment by pb-cdunn** [comment]
If that doesn't help, try to compile with `-E`. Then look at the .o file for the preprocessor output. You should be able to see where `H5Cpp.h` actually comes from and whether it loads `H5Location.h` properly.

My guess (totally a hunch): You have a relative path to **HDF5**, but `make` runs in a sub-directory. So your relative path is wrong, and you are picking up HDF5 from the system path. But really, that's just my guess. Try running `configure.py` again with `HDF5_INC=absolute-path`.

**Comment by mhsieh** [comment]
Works for me on centos6.6, I am curious to see if these are related to toolchain or CFLAGS CXXFLAGS.

**Comment by mebbert** [comment]
Bah! It was totally the relative paths for HDF5. I wouldn't have guessed that from the errors. I would have expected an error that it couldn't find files. 

Thanks so much for your help.

Also, I found that rerunning ./configure.py with different options does not always update everything for the make. For example, I reran it with the full paths to HDF5 libs, but trying to make still used the relative path. It updated the defines.mk file appropriately, but it didn't translate to the make process. I opted to nuke the blasr directory and re-clone. Everything worked from there.

Let me know if I need to clarify anything.
